File: apps/backend/app/schemas/driver.py
"""Response DTOs for `/drivers*` (docs/08_API_SPECIFICATION.md — "Drivers").

`Driver` is used for both the list and the single-profile endpoint — docs/08
doesn't describe the profile as carrying more fields than the list entry,
so one schema covers both rather than inventing a distinction that isn't
specified anywhere.
"""

from __future__ import annotations

import uuid
from datetime import date

from app.schemas.base import CamelModel


class Driver(CamelModel):
    id: uuid.UUID
    driver_number: int | None
    full_name: str
    abbreviation: str | None
    nationality: str | None
    date_of_birth: date | None
    team_id: uuid.UUID | None
    team: str | None
    rookie_season: int | None
    world_titles: int | None
    active: bool | None


class SeasonDriverSummary(CamelModel):
    season: int
    wins: int
    podiums: int
    poles: int
    fastest_laps: int
    average_finish: float | None
    average_qualifying: float | None
    consistency_score: float | None
    pit_efficiency: float | None
    race_pace: float | None
    qualifying_pace: float | None


class DriverCareerStatistics(CamelModel):
    driver: str
    seasons_competed: int
    wins: int
    podiums: int
    poles: int
    fastest_laps: int
    average_finish: float | None
    consistency_score: float | None


class DriverLap(CamelModel):
    season: int
    round: int
    race_name: str | None
    session_type: str
    lap_number: float
    lap_time: float | None
    sector_1: float | None
    sector_2: float | None
    sector_3: float | None
    compound: str | None
    tyre_life: float | None
    position: float | None
    pit_in: bool | None
    pit_out: bool | None


class SeasonConsistencyEntry(CamelModel):
    season: int
    consistency_score: float | None
    average_finish: float | None


class DriverConsistency(CamelModel):
    driver: str
    career_consistency_score: float | None
    seasons: list[SeasonConsistencyEntry]


class DriverComparisonEntry(CamelModel):
    driver: str
    wins: int
    podiums: int
    poles: int
    fastest_laps: int
    average_finish: float | None
    average_qualifying: float | None
    consistency_score: float | None
    pit_efficiency: float | None
    race_pace: float | None
    qualifying_pace: float | None


class DriverComparison(CamelModel):
    season: int
    driver_a: DriverComparisonEntry
    driver_b: DriverComparisonEntry

File: packages/shared-types/src/domain/driver.ts
/**
 * `/drivers*` (docs/08_API_SPECIFICATION.md — "Drivers"). Mirrors
 * `apps/backend/app/schemas/driver.py` exactly.
 */

export interface Driver {
  id: string;
  driverNumber: number | null;
  fullName: string;
  abbreviation: string | null;
  nationality: string | null;
  dateOfBirth: string | null;
  teamId: string | null;
  team: string | null;
  rookieSeason: number | null;
  worldTitles: number | null;
  active: boolean | null;
}

export interface DriverCareerStatistics {
  driver: string;
  seasonsCompeted: number;
  wins: number;
  podiums: number;
  poles: number;
  fastestLaps: number;
  averageFinish: number | null;
  consistencyScore: number | null;
}

export interface DriverLap {
  season: number;
  round: number;
  raceName: string | null;
  sessionType: string;
  lapNumber: number;
  lapTime: number | null;
  sector1: number | null;
  sector2: number | null;
  sector3: number | null;
  compound: string | null;
  tyreLife: number | null;
  position: number | null;
  pitIn: boolean | null;
  pitOut: boolean | null;
}

export interface SeasonConsistencyEntry {
  season: number;
  consistencyScore: number | null;
  averageFinish: number | null;
}

export interface DriverConsistency {
  driver: string;
  careerConsistencyScore: number | null;
  seasons: SeasonConsistencyEntry[];
}

export interface DriverComparisonEntry {
  driver: string;
  wins: number;
  podiums: number;
  poles: number;
  fastestLaps: number;
  averageFinish: number | null;
  averageQualifying: number | null;
  consistencyScore: number | null;
  pitEfficiency: number | null;
  racePace: number | null;
  qualifyingPace: number | null;
}

export interface DriverComparison {
  season: number;
  driverA: DriverComparisonEntry;
  driverB: DriverComparisonEntry;
}
